{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/gpfs3/well/band/users/rbx225/GAMCC\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 100x100 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import io\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import gzip\n",
    "import glob\n",
    "import time \n",
    "import json\n",
    "import math\n",
    "import pickle\n",
    "import secrets\n",
    "import multiprocessing\n",
    "import subprocess\n",
    "import resource\n",
    "import pyreadr\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import statsmodels.api as sm\n",
    "import random\n",
    "import itertools\n",
    "import collections\n",
    "import pyreadr\n",
    "import pywfa\n",
    "from collections import Counter\n",
    "from sklearn.cluster import KMeans\n",
    "from scipy.spatial.distance import cdist\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from sklearn.preprocessing import normalize\n",
    "from sklearn.preprocessing import minmax_scale\n",
    "\n",
    "import multiprocessing as mp\n",
    "from functools import partial\n",
    "from itertools import combinations_with_replacement\n",
    "\n",
    "home_dir = os.environ.get(\"home_dir\")\n",
    "sys.path.append(f'{home_dir}software/lcwgsus/')\n",
    "import lcwgsus\n",
    "from lcwgsus.variables import *\n",
    "from warnings import simplefilter\n",
    "from hla_phase_functions import *\n",
    "from hla_align_functions import *\n",
    "\n",
    "from scipy.special import logsumexp\n",
    "from scipy.stats import pearsonr\n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "%cd /well/band/users/rbx225/GAMCC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optimising hyperparameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>prob</th>\n",
       "      <th>cum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A*23:01</td>\n",
       "      <td>A*23:01</td>\n",
       "      <td>0.999968</td>\n",
       "      <td>0.999968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   allele1  allele2      prob       cum\n",
       "0  A*23:01  A*23:01  0.999968  0.999968"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score_diff_in_alignment_genes_ary = [0, 8]\n",
    "n_mismatches_ary = [3, 5]\n",
    "weight_ary = [True, False]\n",
    "\n",
    "d = score_diff_in_alignment_genes_ary[0]\n",
    "n = n_mismatches_ary[0]\n",
    "w = weight_ary[0]\n",
    "\n",
    "g = 'A'\n",
    "s = 'IDT0481'\n",
    "\n",
    "bam = f\"/well/band/users/rbx225/GAMCC/data/bams/{s}.bam\"\n",
    "db_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/v3390_merged_only/'\n",
    "as_dir = 'results/hla/imputation/WFA_alignments/v3390/'\n",
    "hla_gene_information = pd.read_csv(HLA_GENE_INFORMATION_FILE, sep = ' ')\n",
    "\n",
    "r1, _, mate, pair = hla_aligner(g, bam, db_dir, as_dir, hla_gene_information,\n",
    "                               n_mismatches = n, score_diff_in_alignment_genes = d)\n",
    "\n",
    "qmat = pyreadr.read_r(f'results/hla/imputation/QUILT_HLA_result_method/{s}/{g}/extracted.hla{g}.RData')['quilt']\n",
    "\n",
    "a1 = set(pair.columns)\n",
    "a2 = set(qmat.columns)\n",
    "\n",
    "cols = sorted(list(a1.intersection(a2)))\n",
    "pair = pair.loc[cols, cols]\n",
    "qmat = qmat.loc[cols, cols]\n",
    "\n",
    "pair = pair - pair.max().max()\n",
    "mask = np.tril(np.ones(pair.shape), k=-1).astype(bool)\n",
    "pair = pair.where(~mask, other=0)\n",
    "pair = np.exp(pair)\n",
    "pair = pair/pair.sum().sum()\n",
    "\n",
    "if w:\n",
    "    w1 = pair.max().max()\n",
    "    w2 = qmat.max().max()\n",
    "    w1 = w1/(w1+w2)\n",
    "    w2 = 1-w1\n",
    "    \n",
    "    combined = (pair**w1)*(qmat**w2)\n",
    "else:\n",
    "    combined = pair*qmat\n",
    "    \n",
    "combined = combined/combined.sum().sum()\n",
    "bestalleles = get_best_alleles(np.log(combined))\n",
    "\n",
    "bestalleles.columns = ['bestallele1', 'bestallele2', 'post_prob', 'sums']\n",
    "bestalleles['sample_number'] = 1\n",
    "bestalleles['sample_name'] = sample\n",
    "bestalleles = bestalleles[['sample_number', 'sample_name', 'bestallele1', 'bestallele2', 'post_prob', 'sums']]\n",
    "\n",
    "bestalleles.to_csv(f'{reads_df_outdir}/{sample}/{gene}/quilt.hla.output.onlyreads.all.txt', index = False, header = True, sep = '\\t')\n",
    "bestalleles.iloc[[0], :].to_csv(f'{reads_df_outdir}/{sample}/{gene}/quilt.hla.output.onlyreads.topresult.txt', index = False, header = True, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Comparing with QUILT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['B*08:01', 'B*35:12'], dtype='<U7')"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene = 'B'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "bam = \"/well/band/users/rbx225/GAMCC/data/bams/IDT0561.bam\"\n",
    "db_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/v3390_oneKG_only/'\n",
    "reads_apart_max=1000\n",
    "reads_extend_max=1000\n",
    "n_mismatches=5\n",
    "assumed_bq=0.001\n",
    "score_diff_in_alignment_genes = 0\n",
    "n_unique_onefield = 100\n",
    "phase=False\n",
    "\n",
    "reads1 = get_chr6_reads(gene, bam, hla_gene_information, strict = True,\n",
    "                            reads_apart_max = reads_apart_max, \n",
    "                            reads_extend_max = reads_extend_max)\n",
    "    \n",
    "rl = reads1['sequence'].str.len().mode().values[0]\n",
    "\n",
    "columns = []\n",
    "for g in HLA_GENES_ALL:\n",
    "    db = pd.read_csv(f'{db_dir}{g}.ssv', sep = ' ')\n",
    "    columns = columns + db.columns.tolist()\n",
    "columns = np.array(columns)\n",
    "columns_to_keep = np.where(np.char.startswith(columns, gene + '*'))[0]\n",
    "target_alleles = columns[columns_to_keep]\n",
    "sample = bam.split('/')[-1].split('.')[0]\n",
    "\n",
    "likemat1 = pd.read_csv(f\"/well/band/users/rbx225/GAMCC/results/hla/imputation/WFA_alignments/v3390_oneKG/{sample}/{gene}/AS_matrix.ssv\", sep = ' ', index_col = 0).values\n",
    "min_valid_prob = n_mismatches*np.log(assumed_bq) + (rl - n_mismatches)*np.log(1 - assumed_bq)\n",
    "\n",
    "valid_indices1 = np.any(likemat1 >= min_valid_prob, axis=1)\n",
    "likemat1, reads1 = likemat1[valid_indices1], reads1[valid_indices1]\n",
    "likemat_all = likemat1\n",
    "id1 = reads1.iloc[:, 0].to_numpy()\n",
    "\n",
    "unique_ids = np.unique(id1)\n",
    "likemat_mate = -600*np.ones((len(unique_ids), likemat_all.shape[1]))\n",
    "\n",
    "for i, uid in enumerate(unique_ids):\n",
    "    tmp = likemat_all[id1 == uid, :]\n",
    "\n",
    "    keep = False\n",
    "    unique = True\n",
    "    for j in range(tmp.shape[0]):\n",
    "        best_indices = np.where(tmp[j,:] == tmp[j,:].max())[0]\n",
    "        best_alleles = columns[best_indices]\n",
    "        aligned_genes = np.unique(np.array([s.split('*')[0] for s in best_alleles]))\n",
    "        keep = keep or ((len(aligned_genes) == 1) and (aligned_genes[0] == gene))\n",
    "        unique = unique and (gene in aligned_genes)\n",
    "\n",
    "    if (keep and unique):\n",
    "        likemat_mate[i, :] = np.sum(tmp, axis=0)\n",
    "\n",
    "# Filtering out based on alignment scores\n",
    "\n",
    "scoring_df = pd.DataFrame({'ID': unique_ids, \n",
    "                    'Target': likemat_mate[:,columns_to_keep].max(axis = 1), \n",
    "                    'Others': likemat_mate[:,~np.where(np.char.startswith(columns, gene + '*'))[0]].max(axis = 1)\n",
    "                   })\n",
    "scoring_df['diff'] = scoring_df['Target'] - scoring_df['Others']\n",
    "scoring_df\n",
    "valid_mask = ((scoring_df['diff'] > score_diff_in_alignment_genes) & (scoring_df['Target'] >= min_valid_prob)).tolist()\n",
    "\n",
    "likemat_mate = likemat_mate[valid_mask, :][:, columns_to_keep]\n",
    "unique_ids = unique_ids[valid_mask]\n",
    "reads1 = reads1[reads1['ID'].isin(unique_ids)]\n",
    "likemat_mate_df = pd.DataFrame(likemat_mate, index = unique_ids, columns=target_alleles)\n",
    "\n",
    "df = likemat_mate_df.copy()\n",
    "best_alignments = df.max(axis=1)\n",
    "matching_cols = df.eq(best_alignments, axis=0).apply(lambda row: likemat_mate_df.columns[row].tolist(), axis=1)\n",
    "one_field = matching_cols.apply(lambda cols: list(set(col.split(':')[0] for col in cols)))\n",
    "counts = one_field.apply(len)\n",
    "rows_to_remove = counts[counts >= n_unique_onefield].index\n",
    "rows_to_keep = df.index.difference(rows_to_remove)\n",
    "\n",
    "likemat_mate_df = likemat_mate_df.loc[rows_to_keep]\n",
    "reads1 = reads1[reads1['ID'].isin(rows_to_keep)]\n",
    "likemat_mate = likemat_mate_df.values\n",
    "\n",
    "if phase:\n",
    "    if likemat_mate.shape[0] > 2:\n",
    "        likemat_mate_normalised = normalize(np.exp(likemat_mate - likemat_mate.max(axis = 1, keepdims = True)), axis=1, norm='l1')\n",
    "        gmm = GaussianMixture(n_components=2, tol = 1e-10, max_iter = 1000, n_init = 50)\n",
    "        gmm.fit(likemat_mate_normalised)\n",
    "        labels = gmm.predict(likemat_mate_normalised)\n",
    "\n",
    "        group1 = np.where(labels == 0)[0]\n",
    "        group2 = np.where(labels == 1)[0]\n",
    "\n",
    "        l1 = likemat_mate[group1, :].mean(axis=0)\n",
    "        l2 = likemat_mate[group2, :].mean(axis=0)\n",
    "\n",
    "        likemat_mate = np.vstack([l1, l2])\n",
    "\n",
    "likemat_norm=likemat_mate-likemat_mate.max(axis = 1, keepdims = True)\n",
    "likemat_norm=0.5*np.exp(likemat_norm)+1e-100\n",
    "\n",
    "likemat_pair=pd.DataFrame(0, index=target_alleles, columns=target_alleles)\n",
    "qq=likemat_pair*0\n",
    "for i in range(likemat_mate.shape[0]):\n",
    "    qq=qq*0\n",
    "    m1=qq+likemat_norm[i,:]\n",
    "    m1=m1+m1.T\n",
    "    likemat_pair=likemat_pair+np.log(m1)\n",
    "\n",
    "likemat_paired_df = pd.DataFrame(likemat_pair, index=target_alleles, columns=target_alleles)\n",
    "lcwgsus.extract_unique_twofield(find_best_pair_alignment(likemat_paired_df)[0].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['B*07:02' 'B*08:01' 'B*40:01' 'B*40:03' 'B*41:01' 'B*41:02' 'B*41:03'\n",
      " 'B*42:01' 'B*42:02' 'B*42:08' 'B*44:03']\n",
      "['B*35:01' 'B*35:02' 'B*35:03' 'B*35:04' 'B*35:05' 'B*35:06' 'B*35:08'\n",
      " 'B*35:09' 'B*35:10' 'B*35:11' 'B*35:12' 'B*35:14' 'B*35:17' 'B*35:20'\n",
      " 'B*35:22' 'B*35:23' 'B*35:40N' 'B*35:41' 'B*35:43' 'B*35:57' 'B*51:01'\n",
      " 'B*51:02' 'B*51:06' 'B*51:07' 'B*51:08' 'B*51:09' 'B*51:83' 'B*52:01'\n",
      " 'B*52:02' 'B*53:01' 'B*58:01' 'B*58:02' 'B*58:11' 'B*78:01' 'B*78:04']\n",
      "['B*08:01']\n",
      "['B*35:01' 'B*35:02' 'B*35:03' 'B*35:05' 'B*35:12' 'B*35:14' 'B*35:17'\n",
      " 'B*58:02']\n",
      "['B*07:02' 'B*08:01' 'B*27:07' 'B*39:01' 'B*40:02' 'B*40:03' 'B*40:04'\n",
      " 'B*40:05' 'B*40:06' 'B*41:01' 'B*41:02' 'B*42:01' 'B*42:02' 'B*48:01']\n"
     ]
    }
   ],
   "source": [
    "for i in range(likemat_mate.shape[0]):\n",
    "    best = target_alleles[np.where(likemat_mate[i,:] == likemat_mate[i,:].max())[0]]\n",
    "    print(lcwgsus.extract_unique_twofield(best))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## BWA alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def pos_to_HLA_allele(r, index_df):\n",
    "    gene, pos = r[['gene', 'pos']]\n",
    "    tmp = index_df[index_df['gene'] == gene]\n",
    "    tmp = tmp[(tmp['Start'] <= pos) & (tmp['End'] >= pos)]\n",
    "    if len(tmp) == 0:\n",
    "        r['pos'] = 'N/A'\n",
    "    else:\n",
    "        r['pos'] = tmp['Allele'].values[0]\n",
    "    \n",
    "    return r\n",
    "\n",
    "def map_alleles_to_array(target_alleles, oneread, misalignment_score = -151):\n",
    "    allele_to_as = dict(zip(oneread['pos'], oneread['AS']))\n",
    "    result = np.array([allele_to_as.get(a, misalignment_score) for a in target_alleles])\n",
    "    return result\n",
    "\n",
    "def get_as_from_one_read_pair(rid, onepair, gene, index_df, target_alleles):\n",
    "    scores_ary_list = []\n",
    "\n",
    "    unique = False\n",
    "    aligned = True\n",
    "\n",
    "    for i in [1,2]:\n",
    "        oneread = onepair[onepair['pair'] == i]\n",
    "        if len(oneread) == 0:\n",
    "            pass\n",
    "        else:\n",
    "            aligned_genes = oneread['gene'].unique()\n",
    "            if gene in aligned_genes:\n",
    "                target_max = oneread[oneread['gene'] == gene]['AS'].max()\n",
    "            else:\n",
    "                target_max = 0\n",
    "\n",
    "            others = oneread[oneread['gene'] != gene]\n",
    "            if len(others) != 0:\n",
    "                others_max = others['AS'].max()\n",
    "            else:\n",
    "                others_max = 0\n",
    "\n",
    "            unique = unique or ((len(aligned_genes) == 1) and (aligned_genes[0] == gene))\n",
    "            aligned = aligned and (gene in alinged_genes)\n",
    "\n",
    "            oneread = oneread[oneread['gene'] == gene]\n",
    "            oneread = oneread.apply(pos_to_HLA_allele, args = (index_df,), axis = 1)\n",
    "            scores_ary = map_alleles_to_array(target_alleles, oneread)\n",
    "            scores_ary_list.append(scores_ary)\n",
    "\n",
    "    if (unique and aligned):\n",
    "        final_scores_ary = np.zeros(len(scores_ary_list)[0])\n",
    "        for s_ary in scores_ary_list:\n",
    "            final_scores_ary = final_scores_ary + s_ary\n",
    "        return rid, final_scores_ary\n",
    "    else:\n",
    "        return '', None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true,
    "hidden": true
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[28], line 19\u001b[0m\n\u001b[1;32m     16\u001b[0m command \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msamtools view \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mbam\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     17\u001b[0m reads \u001b[38;5;241m=\u001b[39m subprocess\u001b[38;5;241m.\u001b[39mrun(command, shell \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m, capture_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m, text \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m)\u001b[38;5;241m.\u001b[39mstdout[:\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 19\u001b[0m reads \u001b[38;5;241m=\u001b[39m [i\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m reads \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m##\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m i]\n\u001b[1;32m     20\u001b[0m reads \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(reads)\n\u001b[1;32m     21\u001b[0m reads\u001b[38;5;241m.\u001b[39mcolumns \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     22\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mID\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mflag\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchr\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpos\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmap_quality\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCIGAR\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchr_alt\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpos_alt\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minsert_size\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msequence\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbase_quality\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     23\u001b[0m ] \u001b[38;5;241m+\u001b[39m [\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcol\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m11\u001b[39m, reads\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m])]\n",
      "Cell \u001b[0;32mIn[28], line 19\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m     16\u001b[0m command \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msamtools view \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mbam\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     17\u001b[0m reads \u001b[38;5;241m=\u001b[39m subprocess\u001b[38;5;241m.\u001b[39mrun(command, shell \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m, capture_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m, text \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m)\u001b[38;5;241m.\u001b[39mstdout[:\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 19\u001b[0m reads \u001b[38;5;241m=\u001b[39m [\u001b[43mi\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msplit\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m reads \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m##\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m i]\n\u001b[1;32m     20\u001b[0m reads \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(reads)\n\u001b[1;32m     21\u001b[0m reads\u001b[38;5;241m.\u001b[39mcolumns \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     22\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mID\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mflag\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchr\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpos\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmap_quality\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCIGAR\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchr_alt\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpos_alt\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minsert_size\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msequence\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbase_quality\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     23\u001b[0m ] \u001b[38;5;241m+\u001b[39m [\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcol\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m11\u001b[39m, reads\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m])]\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "n_mismatch = 5\n",
    "min_valid_score = 151 - 5*n_mismatch\n",
    "gene = 'B'\n",
    "\n",
    "index_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/fasta/v3390_oneKG/'\n",
    "files = glob.glob(f'{index_dir}*.index.tsv')\n",
    "index_df = []\n",
    "for f in files:\n",
    "    tmp = pd.read_csv(f, sep = '\\t')\n",
    "    tmp['gene'] = tmp['Allele'].str.split('*').str.get(0)\n",
    "    index_df.append(tmp)\n",
    "index_df = pd.concat(index_df).reset_index(drop = True)\n",
    "target_alleles = index_df[index_df['gene'] == gene]['Allele'].to_numpy()\n",
    "\n",
    "bam = 'data/realigned_bams/v3390_oneKG/IDT0717.bam'\n",
    "command = f\"samtools view {bam}\"\n",
    "reads = subprocess.run(command, shell = True, capture_output = True, text = True).stdout[:-1].split('\\n')\n",
    "\n",
    "reads = [i.split('\\t') for i in reads if '##' not in i]\n",
    "reads = pd.DataFrame(reads)\n",
    "reads.columns = [\n",
    "    \"ID\", \"flag\", \"chr\", \"pos\", \"map_quality\", \"CIGAR\", \"chr_alt\", \"pos_alt\", \"insert_size\", \"sequence\", \"base_quality\"\n",
    "] + [f'col{i}' for i in range(11, reads.shape[1])]\n",
    "\n",
    "reads['pos'] = reads['pos'].astype(int)\n",
    "reads['pos_alt'] = reads['pos_alt'].astype(int)\n",
    "\n",
    "reads = reads[reads['CIGAR'] != '*']\n",
    "reads['NM'] = reads['col11'].str.split(':').str.get(2).astype(int)\n",
    "reads['AS'] = reads['col13'].str.split(':').str.get(2).astype(int)\n",
    "reads['gene'] = reads['chr'].str.split('-').str.get(1)\n",
    "reads['pair'] = reads['ID'].str.split('_').str.get(1).astype(int)\n",
    "reads['ID'] = reads['ID'].str.split('_').str.get(0)\n",
    "reads = reads[['ID', 'pair', 'gene', 'pos', 'AS', 'NM']]\n",
    "reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>pair</th>\n",
       "      <th>gene</th>\n",
       "      <th>pos</th>\n",
       "      <th>AS</th>\n",
       "      <th>NM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1115:13241:11898</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>90</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1475:4924:10989</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>90</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1573:2582:37027</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2242:7808:4586</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>80</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A00711:748:HHFNVDSX7:4:1602:26603:32847</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>13</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2248794</th>\n",
       "      <td>A00711:748:HHFNVDSX7:4:2419:18285:1251</td>\n",
       "      <td>2</td>\n",
       "      <td>Y</td>\n",
       "      <td>7788</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2248795</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2649:24659:5776</td>\n",
       "      <td>1</td>\n",
       "      <td>Y</td>\n",
       "      <td>7810</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2248796</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1418:29875:36229</td>\n",
       "      <td>1</td>\n",
       "      <td>Y</td>\n",
       "      <td>7867</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2248797</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1418:29875:36229</td>\n",
       "      <td>2</td>\n",
       "      <td>Y</td>\n",
       "      <td>7970</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2248798</th>\n",
       "      <td>A00711:748:HHFNVDSX7:4:2419:18285:1251</td>\n",
       "      <td>1</td>\n",
       "      <td>Y</td>\n",
       "      <td>8027</td>\n",
       "      <td>142</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2248799 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              ID  pair gene   pos   AS  NM\n",
       "0        A00711:753:HHCG3DSX7:4:1115:13241:11898     2    A    11   90   5\n",
       "1         A00711:753:HHCG3DSX7:4:1475:4924:10989     2    A    11   90   5\n",
       "2         A00711:753:HHCG3DSX7:4:1573:2582:37027     2    A    11   79   0\n",
       "3          A00711:753:HHCG3DSX7:4:2242:7808:4586     2    A    11   80   4\n",
       "4        A00711:748:HHFNVDSX7:4:1602:26603:32847     1    A    13   55   0\n",
       "...                                          ...   ...  ...   ...  ...  ..\n",
       "2248794   A00711:748:HHFNVDSX7:4:2419:18285:1251     2    Y  7788  151   0\n",
       "2248795   A00711:753:HHCG3DSX7:4:2649:24659:5776     1    Y  7810  151   0\n",
       "2248796  A00711:753:HHCG3DSX7:4:1418:29875:36229     1    Y  7867  151   0\n",
       "2248797  A00711:753:HHCG3DSX7:4:1418:29875:36229     2    Y  7970  151   0\n",
       "2248798   A00711:748:HHFNVDSX7:4:2419:18285:1251     1    Y  8027  142   0\n",
       "\n",
       "[2248799 rows x 6 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>pair</th>\n",
       "      <th>gene</th>\n",
       "      <th>pos</th>\n",
       "      <th>AS</th>\n",
       "      <th>NM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15603</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2354:20817:19601</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>1293394</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16083</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2354:20817:19601</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>1622564</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16090</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2354:20817:19601</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>1626189</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16092</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2354:20817:19601</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>1629661</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16098</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2354:20817:19601</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>1633683</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23528</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2354:20817:19601</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>4874670</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23535</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2354:20817:19601</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>4877695</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23543</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2354:20817:19601</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>4880759</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23555</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2354:20817:19601</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>4884386</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23652</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2354:20817:19601</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>4917427</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>500 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            ID  pair gene      pos   AS  NM\n",
       "15603  A00711:753:HHCG3DSX7:4:2354:20817:19601     1    B  1293394  151   0\n",
       "16083  A00711:753:HHCG3DSX7:4:2354:20817:19601     1    B  1622564  151   0\n",
       "16090  A00711:753:HHCG3DSX7:4:2354:20817:19601     1    B  1626189  151   0\n",
       "16092  A00711:753:HHCG3DSX7:4:2354:20817:19601     1    B  1629661  151   0\n",
       "16098  A00711:753:HHCG3DSX7:4:2354:20817:19601     1    B  1633683  151   0\n",
       "...                                        ...   ...  ...      ...  ...  ..\n",
       "23528  A00711:753:HHCG3DSX7:4:2354:20817:19601     1    B  4874670  151   0\n",
       "23535  A00711:753:HHCG3DSX7:4:2354:20817:19601     1    B  4877695  151   0\n",
       "23543  A00711:753:HHCG3DSX7:4:2354:20817:19601     1    B  4880759  151   0\n",
       "23555  A00711:753:HHCG3DSX7:4:2354:20817:19601     1    B  4884386  151   0\n",
       "23652  A00711:753:HHCG3DSX7:4:2354:20817:19601     1    B  4917427  151   0\n",
       "\n",
       "[500 rows x 6 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "onepair[onepair['pair'] == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'alinged_genes' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[55], line 30\u001b[0m\n\u001b[1;32m     27\u001b[0m     others_max \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m     29\u001b[0m unique \u001b[38;5;241m=\u001b[39m unique \u001b[38;5;129;01mor\u001b[39;00m ((\u001b[38;5;28mlen\u001b[39m(aligned_genes) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m (aligned_genes[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m==\u001b[39m gene))\n\u001b[0;32m---> 30\u001b[0m aligned \u001b[38;5;241m=\u001b[39m aligned \u001b[38;5;129;01mand\u001b[39;00m (gene \u001b[38;5;129;01min\u001b[39;00m \u001b[43malinged_genes\u001b[49m)\n\u001b[1;32m     32\u001b[0m oneread \u001b[38;5;241m=\u001b[39m oneread[oneread[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m gene]\n\u001b[1;32m     33\u001b[0m oneread \u001b[38;5;241m=\u001b[39m oneread\u001b[38;5;241m.\u001b[39mapply(pos_to_HLA_allele, args \u001b[38;5;241m=\u001b[39m (index_df,), axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'alinged_genes' is not defined"
     ]
    }
   ],
   "source": [
    "rid = 'A00711:753:HHCG3DSX7:4:2354:20817:19601'\n",
    "onepair = reads[reads['ID'] == rid]\n",
    "gene = 'B'\n",
    "index_df = index_df\n",
    "target_alleles = target_alleles\n",
    "\n",
    "scores_ary_list = []\n",
    "\n",
    "unique = False\n",
    "aligned = True\n",
    "\n",
    "for i in [1,2]:\n",
    "    oneread = onepair[onepair['pair'] == i]\n",
    "    if len(oneread) == 0:\n",
    "        pass\n",
    "    else:\n",
    "        aligned_genes = oneread['gene'].unique()\n",
    "        if gene in aligned_genes:\n",
    "            target_max = oneread[oneread['gene'] == gene]['AS'].max()\n",
    "        else:\n",
    "            target_max = 0\n",
    "\n",
    "        others = oneread[oneread['gene'] != gene]\n",
    "        if len(others) != 0:\n",
    "            others_max = others['AS'].max()\n",
    "        else:\n",
    "            others_max = 0\n",
    "\n",
    "        unique = unique or ((len(aligned_genes) == 1) and (aligned_genes[0] == gene))\n",
    "        aligned = aligned and (gene in alinged_genes)\n",
    "\n",
    "        oneread = oneread[oneread['gene'] == gene]\n",
    "        oneread = oneread.apply(pos_to_HLA_allele, args = (index_df,), axis = 1)\n",
    "        scores_ary = map_alleles_to_array(target_alleles, oneread)\n",
    "        scores_ary_list.append(scores_ary)\n",
    "scores_ary_list\n",
    "# if (unique and aligned):\n",
    "#     final_scores_ary = np.zeros(len(scores_ary_list)[0])\n",
    "#     for s_ary in scores_ary_list:\n",
    "#         final_scores_ary = final_scores_ary + s_ary\n",
    "#     return rid, final_scores_ary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_ids = []\n",
    "\n",
    "rid = 'A00711:753:HHCG3DSX7:4:2354:20817:19601'\n",
    "onepair = reads[reads['ID'] == rid]\n",
    "gene = 'B'\n",
    "index_df = index_df\n",
    "target_alleles = target_alleles\n",
    "\n",
    "\n",
    "    \n",
    "rid, final_scores_ary = get_as_from_one_read_pair(rid, onepair, gene, index_df, target_alleles)\n",
    "rid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Examining Individual cases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IDT0486-A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'A'\n",
    "bam = \"/well/band/users/rbx225/GAMCC/data/bams/IDT0606.bam\"\n",
    "db_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/v3390_merged_only/'\n",
    "as_dir = 'results/hla/imputation/WFA_alignments/v3390/'\n",
    "hla_gene_information = pd.read_csv(HLA_GENE_INFORMATION_FILE, sep = ' ')\n",
    "\n",
    "r1, _, mate, pair = hla_aligner(gene, bam, db_dir, as_dir, hla_gene_information)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A*01:01</th>\n",
       "      <th>A*01:02</th>\n",
       "      <th>A*02:01</th>\n",
       "      <th>A*02:02</th>\n",
       "      <th>A*02:03</th>\n",
       "      <th>A*02:04</th>\n",
       "      <th>A*02:05</th>\n",
       "      <th>A*02:06</th>\n",
       "      <th>A*02:07</th>\n",
       "      <th>A*02:08</th>\n",
       "      <th>...</th>\n",
       "      <th>A*68:01</th>\n",
       "      <th>A*68:02</th>\n",
       "      <th>A*68:03</th>\n",
       "      <th>A*68:15</th>\n",
       "      <th>A*68:17</th>\n",
       "      <th>A*69:01</th>\n",
       "      <th>A*74:01</th>\n",
       "      <th>A*74:02</th>\n",
       "      <th>A*74:03</th>\n",
       "      <th>A*80:01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A*01:01</th>\n",
       "      <td>-141.99005</td>\n",
       "      <td>-149.331113</td>\n",
       "      <td>-124.092316</td>\n",
       "      <td>-136.088788</td>\n",
       "      <td>-132.090182</td>\n",
       "      <td>-132.090182</td>\n",
       "      <td>-136.089118</td>\n",
       "      <td>-132.108015</td>\n",
       "      <td>-132.090182</td>\n",
       "      <td>-136.089124</td>\n",
       "      <td>...</td>\n",
       "      <td>-128.109415</td>\n",
       "      <td>-132.108351</td>\n",
       "      <td>-124.110150</td>\n",
       "      <td>-132.108351</td>\n",
       "      <td>-128.109415</td>\n",
       "      <td>-124.110480</td>\n",
       "      <td>-136.764101</td>\n",
       "      <td>-136.764056</td>\n",
       "      <td>-140.088060</td>\n",
       "      <td>-144.780796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*01:02</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>-191.478938</td>\n",
       "      <td>-128.109421</td>\n",
       "      <td>-148.085931</td>\n",
       "      <td>-140.105892</td>\n",
       "      <td>-140.105892</td>\n",
       "      <td>-151.409889</td>\n",
       "      <td>-144.086995</td>\n",
       "      <td>-140.105892</td>\n",
       "      <td>-152.084867</td>\n",
       "      <td>...</td>\n",
       "      <td>-143.412017</td>\n",
       "      <td>-148.085931</td>\n",
       "      <td>-136.089125</td>\n",
       "      <td>-148.085931</td>\n",
       "      <td>-143.412017</td>\n",
       "      <td>-139.413082</td>\n",
       "      <td>-150.779345</td>\n",
       "      <td>-146.780409</td>\n",
       "      <td>-158.777209</td>\n",
       "      <td>-164.082340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*02:01</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-202.698012</td>\n",
       "      <td>-201.390380</td>\n",
       "      <td>-209.291757</td>\n",
       "      <td>-209.291757</td>\n",
       "      <td>-205.985968</td>\n",
       "      <td>-209.292087</td>\n",
       "      <td>-209.291757</td>\n",
       "      <td>-205.985968</td>\n",
       "      <td>...</td>\n",
       "      <td>-208.617115</td>\n",
       "      <td>-209.292093</td>\n",
       "      <td>-204.694075</td>\n",
       "      <td>-209.292093</td>\n",
       "      <td>-208.617115</td>\n",
       "      <td>-205.293158</td>\n",
       "      <td>-204.084308</td>\n",
       "      <td>-208.065409</td>\n",
       "      <td>-208.029072</td>\n",
       "      <td>-156.744182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*02:02</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-219.996677</td>\n",
       "      <td>-213.351191</td>\n",
       "      <td>-213.351191</td>\n",
       "      <td>-220.671655</td>\n",
       "      <td>-216.675149</td>\n",
       "      <td>-211.386659</td>\n",
       "      <td>-220.689488</td>\n",
       "      <td>...</td>\n",
       "      <td>-213.369024</td>\n",
       "      <td>-217.367960</td>\n",
       "      <td>-208.695442</td>\n",
       "      <td>-217.367960</td>\n",
       "      <td>-213.369024</td>\n",
       "      <td>-209.370419</td>\n",
       "      <td>-212.757442</td>\n",
       "      <td>-214.771005</td>\n",
       "      <td>-218.769658</td>\n",
       "      <td>-172.721756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*02:03</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-219.920775</td>\n",
       "      <td>-219.920775</td>\n",
       "      <td>-217.964606</td>\n",
       "      <td>-218.658090</td>\n",
       "      <td>-214.674670</td>\n",
       "      <td>-217.964936</td>\n",
       "      <td>...</td>\n",
       "      <td>-217.289628</td>\n",
       "      <td>-220.613916</td>\n",
       "      <td>-213.290693</td>\n",
       "      <td>-220.613916</td>\n",
       "      <td>-217.289628</td>\n",
       "      <td>-213.308526</td>\n",
       "      <td>-208.776054</td>\n",
       "      <td>-212.757111</td>\n",
       "      <td>-216.026944</td>\n",
       "      <td>-168.065675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*69:01</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-219.920777</td>\n",
       "      <td>-200.778520</td>\n",
       "      <td>-204.759622</td>\n",
       "      <td>-208.047236</td>\n",
       "      <td>-168.065677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*74:01</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-259.959064</td>\n",
       "      <td>-256.652940</td>\n",
       "      <td>-261.309021</td>\n",
       "      <td>-168.759159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*74:02</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-259.959063</td>\n",
       "      <td>-259.940558</td>\n",
       "      <td>-168.759115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*74:03</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-269.894599</td>\n",
       "      <td>-172.758090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*80:01</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1003.444795</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 74 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           A*01:01     A*01:02     A*02:01     A*02:02     A*02:03  \\\n",
       "A*01:01 -141.99005 -149.331113 -124.092316 -136.088788 -132.090182   \n",
       "A*01:02    0.00000 -191.478938 -128.109421 -148.085931 -140.105892   \n",
       "A*02:01    0.00000    0.000000 -202.698012 -201.390380 -209.291757   \n",
       "A*02:02    0.00000    0.000000    0.000000 -219.996677 -213.351191   \n",
       "A*02:03    0.00000    0.000000    0.000000    0.000000 -219.920775   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "A*69:01    0.00000    0.000000    0.000000    0.000000    0.000000   \n",
       "A*74:01    0.00000    0.000000    0.000000    0.000000    0.000000   \n",
       "A*74:02    0.00000    0.000000    0.000000    0.000000    0.000000   \n",
       "A*74:03    0.00000    0.000000    0.000000    0.000000    0.000000   \n",
       "A*80:01    0.00000    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "            A*02:04     A*02:05     A*02:06     A*02:07     A*02:08  ...  \\\n",
       "A*01:01 -132.090182 -136.089118 -132.108015 -132.090182 -136.089124  ...   \n",
       "A*01:02 -140.105892 -151.409889 -144.086995 -140.105892 -152.084867  ...   \n",
       "A*02:01 -209.291757 -205.985968 -209.292087 -209.291757 -205.985968  ...   \n",
       "A*02:02 -213.351191 -220.671655 -216.675149 -211.386659 -220.689488  ...   \n",
       "A*02:03 -219.920775 -217.964606 -218.658090 -214.674670 -217.964936  ...   \n",
       "...             ...         ...         ...         ...         ...  ...   \n",
       "A*69:01    0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "A*74:01    0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "A*74:02    0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "A*74:03    0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "A*80:01    0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "\n",
       "            A*68:01     A*68:02     A*68:03     A*68:15     A*68:17  \\\n",
       "A*01:01 -128.109415 -132.108351 -124.110150 -132.108351 -128.109415   \n",
       "A*01:02 -143.412017 -148.085931 -136.089125 -148.085931 -143.412017   \n",
       "A*02:01 -208.617115 -209.292093 -204.694075 -209.292093 -208.617115   \n",
       "A*02:02 -213.369024 -217.367960 -208.695442 -217.367960 -213.369024   \n",
       "A*02:03 -217.289628 -220.613916 -213.290693 -220.613916 -217.289628   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "A*69:01    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "A*74:01    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "A*74:02    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "A*74:03    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "A*80:01    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "            A*69:01     A*74:01     A*74:02     A*74:03      A*80:01  \n",
       "A*01:01 -124.110480 -136.764101 -136.764056 -140.088060  -144.780796  \n",
       "A*01:02 -139.413082 -150.779345 -146.780409 -158.777209  -164.082340  \n",
       "A*02:01 -205.293158 -204.084308 -208.065409 -208.029072  -156.744182  \n",
       "A*02:02 -209.370419 -212.757442 -214.771005 -218.769658  -172.721756  \n",
       "A*02:03 -213.308526 -208.776054 -212.757111 -216.026944  -168.065675  \n",
       "...             ...         ...         ...         ...          ...  \n",
       "A*69:01 -219.920777 -200.778520 -204.759622 -208.047236  -168.065677  \n",
       "A*74:01    0.000000 -259.959064 -256.652940 -261.309021  -168.759159  \n",
       "A*74:02    0.000000    0.000000 -259.959063 -259.940558  -168.759115  \n",
       "A*74:03    0.000000    0.000000    0.000000 -269.894599  -172.758090  \n",
       "A*80:01    0.000000    0.000000    0.000000    0.000000 -1003.444795  \n",
       "\n",
       "[74 rows x 74 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.04651972753484129\n",
      "0.22923436030194602\n",
      "0.6840473683209383\n"
     ]
    }
   ],
   "source": [
    "a1 = 'A*01:01'\n",
    "a2 = 'A*36:01'\n",
    "\n",
    "print(x.loc[a1, a2])\n",
    "print(y.loc[a1, a2])\n",
    "print(z.loc[a1, a2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0011079174528854001\n",
      "0.450213985456336\n",
      "0.1936214175202403\n"
     ]
    }
   ],
   "source": [
    "a1 = 'A*01:01'\n",
    "a2 = 'A*01:01'\n",
    "\n",
    "print(x.loc[a1, a2])\n",
    "print(y.loc[a1, a2])\n",
    "print(z.loc[a1, a2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "ename": "PyreadrError",
     "evalue": "File b'results/hla/imputation/QUILT_HLA_result_method/IDT0606/A/test.RData' does not exist!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPyreadrError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[27], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m data \u001b[38;5;241m=\u001b[39m \u001b[43mpyreadr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_r\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mresults/hla/imputation/QUILT_HLA_result_method/IDT0606/A/test.RData\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m x,y,z \u001b[38;5;241m=\u001b[39m data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mx\u001b[39m\u001b[38;5;124m'\u001b[39m], data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124my\u001b[39m\u001b[38;5;124m'\u001b[39m], data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mz\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# x = x.loc[y.columns, y.columns]\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# x = x/x.sum().sum()\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[38;5;66;03m# z = (x**w1)*(y**w2)\u001b[39;00m\n\u001b[1;32m     13\u001b[0m \u001b[38;5;66;03m# z = z/z.sum().sum()\u001b[39;00m\n",
      "File \u001b[0;32m/well/band/users/rbx225/conda/skylake/envs/sus/lib/python3.8/site-packages/pyreadr/pyreadr.py:65\u001b[0m, in \u001b[0;36mread_r\u001b[0;34m(path, use_objects, timezone)\u001b[0m\n\u001b[1;32m     63\u001b[0m filename_bytes \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mexpanduser(filename_bytes)\n\u001b[1;32m     64\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39misfile(filename_bytes):\n\u001b[0;32m---> 65\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PyreadrError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFile \u001b[39m\u001b[38;5;132;01m{0}\u001b[39;00m\u001b[38;5;124m does not exist!\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(filename_bytes))\n\u001b[1;32m     66\u001b[0m parser\u001b[38;5;241m.\u001b[39mparse(filename_bytes)\n\u001b[1;32m     68\u001b[0m result \u001b[38;5;241m=\u001b[39m OrderedDict()\n",
      "\u001b[0;31mPyreadrError\u001b[0m: File b'results/hla/imputation/QUILT_HLA_result_method/IDT0606/A/test.RData' does not exist!"
     ]
    }
   ],
   "source": [
    "data = pyreadr.read_r('results/hla/imputation/QUILT_HLA_result_method/IDT0606/A/test.RData')\n",
    "x,y,z = data['x'], data['y'], data['z']\n",
    "\n",
    "# x = x.loc[y.columns, y.columns]\n",
    "# x = x/x.sum().sum()\n",
    "\n",
    "# w1 = x.max().max()\n",
    "# w2 = y.max().max()\n",
    "# w1 = w1/(w1+w2)\n",
    "# w2 = 1-w1\n",
    "\n",
    "# z = (x**w1)*(y**w2)\n",
    "# z = z/z.sum().sum()\n",
    "z\n",
    "    \n",
    "# get_best_alleles(np.log(z))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>prob</th>\n",
       "      <th>cum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A*01:01</td>\n",
       "      <td>A*36:01</td>\n",
       "      <td>0.280366</td>\n",
       "      <td>0.280366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A*01:01</td>\n",
       "      <td>A*01:01</td>\n",
       "      <td>0.274823</td>\n",
       "      <td>0.555189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A*01:01</td>\n",
       "      <td>A*01:02</td>\n",
       "      <td>0.105761</td>\n",
       "      <td>0.660950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A*01:01</td>\n",
       "      <td>A*30:01</td>\n",
       "      <td>0.059779</td>\n",
       "      <td>0.720729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A*01:02</td>\n",
       "      <td>A*36:01</td>\n",
       "      <td>0.049311</td>\n",
       "      <td>0.770040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>A*01:01</td>\n",
       "      <td>A*31:12</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.989102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>A*01:01</td>\n",
       "      <td>A*31:16</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.989362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>A*01:02</td>\n",
       "      <td>A*03:02</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.989612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>A*24:20</td>\n",
       "      <td>A*36:01</td>\n",
       "      <td>0.000235</td>\n",
       "      <td>0.989846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>A*01:02</td>\n",
       "      <td>A*26:01</td>\n",
       "      <td>0.000227</td>\n",
       "      <td>0.990073</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>134 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     allele1  allele2      prob       cum\n",
       "0    A*01:01  A*36:01  0.280366  0.280366\n",
       "1    A*01:01  A*01:01  0.274823  0.555189\n",
       "2    A*01:01  A*01:02  0.105761  0.660950\n",
       "3    A*01:01  A*30:01  0.059779  0.720729\n",
       "4    A*01:02  A*36:01  0.049311  0.770040\n",
       "..       ...      ...       ...       ...\n",
       "129  A*01:01  A*31:12  0.000260  0.989102\n",
       "130  A*01:01  A*31:16  0.000260  0.989362\n",
       "131  A*01:02  A*03:02  0.000249  0.989612\n",
       "132  A*24:20  A*36:01  0.000235  0.989846\n",
       "133  A*01:02  A*26:01  0.000227  0.990073\n",
       "\n",
       "[134 rows x 4 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pyreadr.read_r('results/hla/imputation/QUILT_HLA_result_method/IDT0606/A/test.RData')\n",
    "x,y = data['x'], data['y']\n",
    "\n",
    "z = (x**x.max().max())*(y**y.max().max())\n",
    "z = z/z.sum().sum()\n",
    "    \n",
    "get_best_alleles(np.log(z))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A*01:01:01:11</th>\n",
       "      <th>A*36:01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A00711:748:HHFNVDSX7:4:1230:18710:27226</th>\n",
       "      <td>-8.515172</td>\n",
       "      <td>-8.515172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:748:HHFNVDSX7:4:1646:14940:22952</th>\n",
       "      <td>-1.251082</td>\n",
       "      <td>-1.251082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:1125:31448:8422</th>\n",
       "      <td>-0.156067</td>\n",
       "      <td>-0.156067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:1237:17517:18959</th>\n",
       "      <td>-0.132365</td>\n",
       "      <td>-0.132365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:1257:9796:34460</th>\n",
       "      <td>-0.640918</td>\n",
       "      <td>-0.640918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:1471:17219:28119</th>\n",
       "      <td>-0.629067</td>\n",
       "      <td>-0.629067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:1474:4887:33317</th>\n",
       "      <td>-0.312134</td>\n",
       "      <td>-0.312134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:1620:3233:22091</th>\n",
       "      <td>-8.150836</td>\n",
       "      <td>-8.150836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:2102:22101:20697</th>\n",
       "      <td>-0.861039</td>\n",
       "      <td>-0.861039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:2220:20455:13369</th>\n",
       "      <td>-37.176296</td>\n",
       "      <td>-37.176296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:2267:11532:7498</th>\n",
       "      <td>-5.297422</td>\n",
       "      <td>-9.296358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:2560:16875:25488</th>\n",
       "      <td>-0.593514</td>\n",
       "      <td>-0.593514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A00711:753:HHCG3DSX7:4:2604:4056:1846</th>\n",
       "      <td>-0.132365</td>\n",
       "      <td>-0.132365</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         A*01:01:01:11    A*36:01\n",
       "A00711:748:HHFNVDSX7:4:1230:18710:27226      -8.515172  -8.515172\n",
       "A00711:748:HHFNVDSX7:4:1646:14940:22952      -1.251082  -1.251082\n",
       "A00711:753:HHCG3DSX7:4:1125:31448:8422       -0.156067  -0.156067\n",
       "A00711:753:HHCG3DSX7:4:1237:17517:18959      -0.132365  -0.132365\n",
       "A00711:753:HHCG3DSX7:4:1257:9796:34460       -0.640918  -0.640918\n",
       "A00711:753:HHCG3DSX7:4:1471:17219:28119      -0.629067  -0.629067\n",
       "A00711:753:HHCG3DSX7:4:1474:4887:33317       -0.312134  -0.312134\n",
       "A00711:753:HHCG3DSX7:4:1620:3233:22091       -8.150836  -8.150836\n",
       "A00711:753:HHCG3DSX7:4:2102:22101:20697      -0.861039  -0.861039\n",
       "A00711:753:HHCG3DSX7:4:2220:20455:13369     -37.176296 -37.176296\n",
       "A00711:753:HHCG3DSX7:4:2267:11532:7498       -5.297422  -9.296358\n",
       "A00711:753:HHCG3DSX7:4:2560:16875:25488      -0.593514  -0.593514\n",
       "A00711:753:HHCG3DSX7:4:2604:4056:1846        -0.132365  -0.132365"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mate[['A*01:01:01:11', 'A*36:01']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# WFA piece"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    score: -6310\n",
       "    pattern_start: 1345\n",
       "    pattern_end: 1496\n",
       "    text_start: 0\n",
       "    text_end: 151\n",
       "    cigartuples: [(0, 151)]\n",
       "    pattern: CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTGGCTCTCAGGGTCTCAGGCCCCGAAGGCGGTGTATGGATTGGGGAGTCCCAGCCTTGGGGATTCCCCAACTCCGCAGTTTCTTTTCTCCCTCTCCCAACCTACGTAGGGTCCTTCATCCTGGATACTCACGACGCGGACCCAGTTCTCACTCCCATTGGGTGTCGGGTTTCCAGAGAAGCCAATCAGTGTCGTCGCGGTCGCTGTTCTAAAGTCCGCACGCACCCACCGGGACTCAGATTCTCCCCAGACGCCGAGGATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGGCCCTGGCCCTGACCCAGACCTGGGCGGGTGAGTGCGGGGTCGGGAGGGAAACCGCCTCTGCGGGGAGAAGCAAGGGGCCCTCCTGGCGGGGGCGCAGGACCGGGGGAGCCGCGCCGGGAGGAGGGTCGGGCAGGTCTCAGCCACTGCTCGCCCCCAGGCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGGCCGCGGGGAGCCCCGCTTCATCGCCGTGGGCTACGTGGACGACACGCAGTTCGTGCGGTTCGACAGCGACGCCGCGAGCCAGAAGATGGAGCCGCGGGCGCCGTGGATAGAGCAGGAGGGGCCGGAGTATTGGGACCAGGAGACACGGAATATGAAGGCCCACTCACAGACTGACCGAGCGAACCTGGGGACCCTGCGCGGCTACTACAACCAGAGCGAGGACGGTGAGTGACCCCGGCCCGGGGCGCAGGTCACGACCCCTCATCCCCCACGGACGGGCCAGGTCGCCCACAGTCTCCGGGTCCGAGATCCACCCCGAAGCCGCGGGACTCCGAGACCCTTGTCCCGGGAGAGGCCCAGGCGCCTTTACCCGGTTTCATTTTCAGTTTAGGCCAAAAATCCCCCCGGGTTGGTCGGGGCGGGGCGGGGCTCGGGGGACTGGGCTGACCGCGGGGTCGGGGCCAGGTTCTCACACCATCCAGATAATGTATGGCTGCGACGTGGGGCCGGACGGGCGCTTCCTCCGCGGGTACCGGCAGGACGCCTACGACGGCAAGGATTACATCGCCCTGAACGAGGACCTGCGCTCTTGGACCGCGGCGGACATGGCAGCTCAGATCACCAAGCGCAAGTGGGAGGCGGTCCATGCGGCGGAGCAGCGGAGAGTCTACCTGGAGGGCCGGTGCGTGGACGGGCTCCGCAGATACCTGGAGAACGGGAAGGAGACGCTGCAGCGCACGGGTACCAGGGGCCACGGGGCGCCTCCCTGATCGCCTATAGATCTCCCGGGCTGGCCTCCCACAAGGAGGGGAGACAATTGGGACCAACACTAGAATATCACCCTCCCTCTGGTCCTGAGGGAGAGGAATCCTCCTGGGTTTCCAGATCCTGTACCAGAGAGTGACTCTGAGGTTCCGCCCTGCTCTCTGACACAATTAAGGGATAAAATCTCTGAAGGAGTGACGGGAAGACGATCCCTCGAATACTGATGAGTGGCTCCCTTTGACACCGGCAGCAGCCTTGGGCCCGTGACTTTTCCTCTCAGGCCTTGTTCTCTGCTTCACACTCAATGTGTGTGGGGGTCTGAGTCCAGCACTTCTGAGTCTCTCAGCCTCCACTCAGGTCAGGACCAGAAGTCGCTGTTCCCTTCTCAGGGAATAGAAGATTATCCCAGGTGCCTGTGTCCAGGCTGGTGTCTGGGTTCTGTGCTCTCTTCCCCATCCCGGGTGTCCTGTCCATTCTCAAGATGGCCACATGCGTGCTGGTGGAGTGTCCCATGACAGATGCAAAATGCCTGAATTTTCTGACTCTTCCCGTCAGACCCCCCCAAGACACATATGACCCACCACCCCATCTCTGACCATGAGGCCACCCTGAGGTGCTGGGCCCTGGGCTTCTACCCTGCGGAGATCACACTGACCTGGCAGCGGGATGGGGAGGACCAGACCCAGGACACGGAGCTCGTGGAGACCAGGCCTGCAGGGGATGGAACCTTCCAGAAGTGGGCGGCTGTGGTGGTGCCTTCTGGAGAGGAGCAGAGATACACCTGCCATGTGCAGCATGAGGGTCTGCCCAAGCCCCTCACCCTGAGATGGGGTAAGGAGGGAGATGGGGGTGTCATGTCTCTTAGGGAAAGCAGGAGCCTCTCTGGAGACCTTTAGCAGGGTCAGGGCCCCTCACCTTCCCCTCTTTTCCCAGAGCTGTCTTCCCAGCCCACCATCCCCATCGTGGGCATCATTGCTGGCCTGGTTCTCCTTGGAGCTGTGATCACTGGAGCTGTGGTCGCTGCCGTGATGTGGAGGAGGAAGAGCTCAGGTGGAGAAGGGGTGAAGGGTGGGGTCTGAGATTTCTTGTCTCACTGAGGGTTCCAAGCCCCAGCTAGAAATGTGCCCTGTCTCATTACTGGGAAGCACCTTCCACAATCATGGGCCGACCCAGCCTGGGCCCTGTGTGCCAGCACTTACTCTTTTGTAAAGCACCTGTTAAAATGAAGGACAGATTTATCACCTTGATTACGGCGGTGATGGGACCTGATCCCAGCAGTCACAAGTCACAGGGGAAGGTCCCTGAGGACAGACCTCAGGAGGGCTATTGGTCCAGGACCCACACCTGCTTTCTTCATGTTTCCTGATCCCGCCCTGGGTCTGCAGTCACACATTTCTGGAAACTTCTCTGGGGTCCAAGACTAGGAGGTTCCTCTAGGACCTTAAGGCCCTGGCTCCTTTCTGGTATCTCACAGGACATTTTCTTCCCACAGATAGAAAAGGAGGGAGTTACACTCAGGCTGCAAGTAAGTATGAAGGAGGCTGATGCCTGAGGTCCTTGGGATATTGTGTTTGGGAGCCCATGGGGGAGCTCACCCACCCCACAATTCCTCCTCTAGCCACATCTTCTGTGGGATCTGACCAGGTTCTGTTTTTGTTCTACCCCAGGCAGTGACAGTGCCCAGGGCTCTGATGTGTCTCTCACAGCTTGTAAAGGTGAGAGCTTGGAGGGCCTGATGTGTGTTGGGTGTTGGGTGGAACAGTGGACACAGCTGTGCTATGGGGTTTCTTTGCGTTGGATGTATTGAGCATGCGATGGGCTGTTTAAGGTGTGACCCCTCACTGTGATGGATATGAATTTGTTCATGAATATTTTTTTCTATAGTGTGAGACAGCTGCCTTGTGTGGGACTGAGAGGCAAGAGTTGTTCCTGCCCTTCCCTTTGTGACTTGAAGAACCCTGACTTTGTTTCTGCAAAGGCACCTGC\n",
       "    text: TCCCACAAGGAGGGGAGACAATTGGGACCAACACTAGAATATCACCCTCCCTCTGGTCCTGAGGGAGAGGAATCCTCCTGGGTTTCCAGATCCTGTACCAGAGAGTGACTCTGAGGTTCCGCCCTGCTCTCTGACACAATTAAGGGATAAA"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db = pd.read_csv('../recyclable_files/hla_reference_files/v3390_oneKG_only/A.ssv', sep = ' ')\n",
    "ix = 21\n",
    "seq = r1.loc[ix,'sequence']\n",
    "bq = r1.loc[ix,'base_quality']\n",
    "\n",
    "# seq = reverse_complement(seq)\n",
    "# bq = bq[::-1]\n",
    "\n",
    "# a = 'DRB1*13:02:01:01'\n",
    "# a = 'DRB1*03:02:01'\n",
    "a = 'A*01:01:01:11'\n",
    "# a = 'A*36:01'\n",
    "\n",
    "refseq = (''.join(db[a].tolist())).replace('.', '').lstrip('*').rstrip('*')\n",
    "ref = pywfa.WavefrontAligner(refseq)\n",
    "result = ref(seq)\n",
    "result = ref(seq,clip_cigar=True, min_aligned_bases_left=10, min_aligned_bases_right=10)\n",
    "cigars_lst = result.cigartuples\n",
    "refstart, refend, readstart, readend = result.pattern_start, result.pattern_end, result.text_start, result.text_end\n",
    "newrefseq, newseq, newbq = recode_sequences(refseq, seq, bq, cigars_lst, refstart, refend, readstart, readend, rl = 151)\n",
    "score = calculate_score_per_alignment(newrefseq, newseq, newbq)\n",
    "# score = adjust_clipping(score, cigars_lst)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGTTCTCTTTGG'"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "newseq[-12:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GGTTCTCCTTGG'"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "newrefseq[-12:]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "92.9972px",
    "width": "166.996px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "223.352px"
   },
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
