{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 100x100 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import io\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import gzip\n",
    "import time\n",
    "import json\n",
    "import pickle\n",
    "import secrets\n",
    "import multiprocessing\n",
    "import subprocess\n",
    "import resource\n",
    "import pandas as pd\n",
    "from scipy.stats import pearsonr\n",
    "# import sqlite3\n",
    "# from PIL import Image\n",
    "import matplotlib.pyplot as plt\n",
    "# from plotnine import *\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import statsmodels.api as sm\n",
    "import random\n",
    "# from collections import Counter\n",
    "# import seaborn as sns\n",
    "# import matplotlib.colors as mcolors\n",
    "# from matplotlib.ticker import FuncFormatter\n",
    "import itertools\n",
    "import collections\n",
    "import pyreadr\n",
    "import psa\n",
    "import pywfa\n",
    "from IPython.display import display_html\n",
    "# import patchworklib as pw\n",
    "sys.path.append('/well/band/users/rbx225/software/lcwgsus/')\n",
    "import lcwgsus\n",
    "from lcwgsus.variables import *\n",
    "from warnings import simplefilter\n",
    "from hla_phase import *\n",
    "\n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "# Genuine logics changed from the main script were marked by string **X** in the original R files.\n",
    "# Simplicity changes were marked by string **Y**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'DRB1'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "samples_to_remove = []\n",
    "\n",
    "gamcc_hla = lcwgsus.read_hla_direct_sequencing(retain = 'fv', unique_two_field = False)\n",
    "gamcc_hla = gamcc_hla[['SampleID', 'Locus', 'Two field1', 'Two field2']].reset_index(drop = True)\n",
    "\n",
    "colnames = ['Sample ID'] + [label for g in HLA_GENES for label in [f'HLA-{g} 1', f'HLA-{g} 2']]\n",
    "hlatypes = pd.DataFrame(columns = colnames)\n",
    "for s in gamcc_hla['SampleID'].unique():\n",
    "    tmp = gamcc_hla[gamcc_hla['SampleID'] == s]\n",
    "    row = [s] + tmp[['Two field1', 'Two field2']].values.ravel().tolist()\n",
    "    hlatypes.loc[len(hlatypes)] = row\n",
    "\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "phased_vcf = \"/well/band/users/rbx225/GAMCC/results/imputation/vcfs/malariaGen_v1_b38/quilt.chr6.vcf.gz\"\n",
    "bam = \"/well/band/users/rbx225/GAMCC/data/bams/IDT0482.bam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA-A</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29942554</td>\n",
       "      <td>29945455</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HLA-B</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31354483</td>\n",
       "      <td>31357158</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31269169</td>\n",
       "      <td>31272071</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HLA-DMA</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32948864</td>\n",
       "      <td>32953036</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HLA-DMB</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32934971</td>\n",
       "      <td>32940807</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HLA-DOA</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33006838</td>\n",
       "      <td>33009536</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HLA-DOB</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32813216</td>\n",
       "      <td>32816951</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-DPA1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33068650</td>\n",
       "      <td>33073570</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-DPB1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33076042</td>\n",
       "      <td>33086238</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-DPB2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33129114</td>\n",
       "      <td>33129113</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HLA-DQA1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32637459</td>\n",
       "      <td>32642764</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HLA-DQA2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32741444</td>\n",
       "      <td>32746394</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HLA-DQB1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32660236</td>\n",
       "      <td>32666607</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HLA-DRA</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32439951</td>\n",
       "      <td>32443910</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32579091</td>\n",
       "      <td>32589742</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32517795</td>\n",
       "      <td>32530224</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HLA-E</td>\n",
       "      <td>chr6</td>\n",
       "      <td>30489532</td>\n",
       "      <td>30492480</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HLA-F</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29723464</td>\n",
       "      <td>29726407</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HLA-G</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29827039</td>\n",
       "      <td>29830382</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HLA-HFE</td>\n",
       "      <td>chr6</td>\n",
       "      <td>26093232</td>\n",
       "      <td>26093231</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HLA-H</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29891080</td>\n",
       "      <td>29891079</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HLA-J</td>\n",
       "      <td>chr6</td>\n",
       "      <td>30009957</td>\n",
       "      <td>30009956</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HLA-L</td>\n",
       "      <td>chr6</td>\n",
       "      <td>30266952</td>\n",
       "      <td>30266951</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HLA-MICA</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31410715</td>\n",
       "      <td>31412431</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HLA-MICB</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31505643</td>\n",
       "      <td>31509909</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HLA-TAP1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32845579</td>\n",
       "      <td>32852497</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HLA-TAP2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32822289</td>\n",
       "      <td>32838233</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HLA-V</td>\n",
       "      <td>chr6</td>\n",
       "      <td>29797808</td>\n",
       "      <td>29797807</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Name   Chr     Start       End  Strand\n",
       "0      HLA-A  chr6  29942554  29945455       1\n",
       "1      HLA-B  chr6  31354483  31357158      -1\n",
       "2      HLA-C  chr6  31269169  31272071      -1\n",
       "3    HLA-DMA  chr6  32948864  32953036      -1\n",
       "4    HLA-DMB  chr6  32934971  32940807      -1\n",
       "5    HLA-DOA  chr6  33006838  33009536      -1\n",
       "6    HLA-DOB  chr6  32813216  32816951      -1\n",
       "7   HLA-DPA1  chr6  33068650  33073570      -1\n",
       "8   HLA-DPB1  chr6  33076042  33086238       1\n",
       "9   HLA-DPB2  chr6  33129114  33129113       1\n",
       "10  HLA-DQA1  chr6  32637459  32642764       1\n",
       "11  HLA-DQA2  chr6  32741444  32746394       1\n",
       "12  HLA-DQB1  chr6  32660236  32666607      -1\n",
       "13   HLA-DRA  chr6  32439951  32443910       1\n",
       "14  HLA-DRB1  chr6  32579091  32589742      -1\n",
       "15  HLA-DRB5  chr6  32517795  32530224      -1\n",
       "16     HLA-E  chr6  30489532  30492480       1\n",
       "17     HLA-F  chr6  29723464  29726407       1\n",
       "18     HLA-G  chr6  29827039  29830382       1\n",
       "19   HLA-HFE  chr6  26093232  26093231       1\n",
       "20     HLA-H  chr6  29891080  29891079       1\n",
       "21     HLA-J  chr6  30009957  30009956       1\n",
       "22     HLA-L  chr6  30266952  30266951       1\n",
       "23  HLA-MICA  chr6  31410715  31412431       1\n",
       "24  HLA-MICB  chr6  31505643  31509909       1\n",
       "25  HLA-TAP1  chr6  32845579  32852497      -1\n",
       "26  HLA-TAP2  chr6  32822289  32838233      -1\n",
       "27     HLA-V  chr6  29797808  29797807       1"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hla_gene_information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14    32579091\n",
       "Name: Start, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hla_gene_information[hla_gene_information['Name'] == f'HLA-{gene}']['Start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>flag</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>map_quality</th>\n",
       "      <th>CIGAR</th>\n",
       "      <th>chr_alt</th>\n",
       "      <th>pos_alt</th>\n",
       "      <th>insert_size</th>\n",
       "      <th>sequence</th>\n",
       "      <th>base_quality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1421:26404:7263</td>\n",
       "      <td>163</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32577822</td>\n",
       "      <td>60</td>\n",
       "      <td>25M1I125M</td>\n",
       "      <td>=</td>\n",
       "      <td>32578071</td>\n",
       "      <td>408</td>\n",
       "      <td>GATAATGGGTTAAAGAAAAATAAAGCCCACTAGATGTAATTTCCTG...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2404:4969:8469</td>\n",
       "      <td>163</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32577885</td>\n",
       "      <td>60</td>\n",
       "      <td>151M</td>\n",
       "      <td>=</td>\n",
       "      <td>32578028</td>\n",
       "      <td>293</td>\n",
       "      <td>ATCTTCACTGAAATACTCATTCCACACATTTTGGTTCTACCTTGTA...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1308:23882:27508</td>\n",
       "      <td>99</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32577918</td>\n",
       "      <td>60</td>\n",
       "      <td>151M</td>\n",
       "      <td>=</td>\n",
       "      <td>32577971</td>\n",
       "      <td>204</td>\n",
       "      <td>GTTCTACCTTGTAATTCCACACCACGAGTCTCATATAAAATGAGAA...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1176:7979:8359</td>\n",
       "      <td>163</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32578034</td>\n",
       "      <td>60</td>\n",
       "      <td>35M2D17M3I96M</td>\n",
       "      <td>=</td>\n",
       "      <td>32578106</td>\n",
       "      <td>226</td>\n",
       "      <td>TATCAACTTAATATATTTCTATAGGATTTATACTCATAGGGTTTAT...</td>\n",
       "      <td>FFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1176:7979:8359</td>\n",
       "      <td>83</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32578106</td>\n",
       "      <td>60</td>\n",
       "      <td>106M3D45M</td>\n",
       "      <td>=</td>\n",
       "      <td>32578034</td>\n",
       "      <td>-226</td>\n",
       "      <td>ACAAGTAACATATAACTAAAAATAAGACATGCATAAAAACAAGTTT...</td>\n",
       "      <td>FFFFFFFF:FFFFFFFFFFFFFFFFFF,FF,FFFFFFFFFFFF:FF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2118:10872:1188</td>\n",
       "      <td>163</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32589728</td>\n",
       "      <td>40</td>\n",
       "      <td>80M1I70M</td>\n",
       "      <td>=</td>\n",
       "      <td>32589878</td>\n",
       "      <td>303</td>\n",
       "      <td>CTTCAGACACACCATGCTGGAGAAGAGGACAGGACCAGGGGCCAGA...</td>\n",
       "      <td>F:FFFFFFFFFFF:,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1566:29197:8093</td>\n",
       "      <td>99</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32589832</td>\n",
       "      <td>60</td>\n",
       "      <td>20S131M</td>\n",
       "      <td>=</td>\n",
       "      <td>32590055</td>\n",
       "      <td>318</td>\n",
       "      <td>CCACCACTCACATCCCAAACATAGGGAGGAAGTTACTAATTTCTTT...</td>\n",
       "      <td>FF,FFFF:FFFFFFFF:::FF:FFFFFFFF:FFFFFF:F:FFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1604:27570:18646</td>\n",
       "      <td>163</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32589859</td>\n",
       "      <td>45</td>\n",
       "      <td>103M2D48M</td>\n",
       "      <td>=</td>\n",
       "      <td>32589997</td>\n",
       "      <td>288</td>\n",
       "      <td>CTCCTGGATTGGGTAATCTCGTGTCGGAGAACCAATCAGCATCTGA...</td>\n",
       "      <td>FFFFFFFFFFFFFF::FFFFFFF:FFFFFFFFFFFF,F:FFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:2118:10872:1188</td>\n",
       "      <td>83</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32589878</td>\n",
       "      <td>60</td>\n",
       "      <td>84M2D67M</td>\n",
       "      <td>=</td>\n",
       "      <td>32589728</td>\n",
       "      <td>-303</td>\n",
       "      <td>CGTGTCGGAGAACCAATCAGCATCTGAGTTCAACAGCATCATCAGT...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>A00711:753:HHCG3DSX7:4:1604:27570:18646</td>\n",
       "      <td>83</td>\n",
       "      <td>chr6</td>\n",
       "      <td>32589997</td>\n",
       "      <td>60</td>\n",
       "      <td>58M1I92M</td>\n",
       "      <td>=</td>\n",
       "      <td>32589859</td>\n",
       "      <td>-288</td>\n",
       "      <td>AGTACTTGAAAGGTTTGAACCAGTTGCATGTAAAACACTTTAATTG...</td>\n",
       "      <td>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          ID flag   chr       pos map_quality  \\\n",
       "47    A00711:753:HHCG3DSX7:4:1421:26404:7263  163  chr6  32577822          60   \n",
       "49     A00711:753:HHCG3DSX7:4:2404:4969:8469  163  chr6  32577885          60   \n",
       "50   A00711:753:HHCG3DSX7:4:1308:23882:27508   99  chr6  32577918          60   \n",
       "53     A00711:753:HHCG3DSX7:4:1176:7979:8359  163  chr6  32578034          60   \n",
       "55     A00711:753:HHCG3DSX7:4:1176:7979:8359   83  chr6  32578106          60   \n",
       "..                                       ...  ...   ...       ...         ...   \n",
       "138   A00711:753:HHCG3DSX7:4:2118:10872:1188  163  chr6  32589728          40   \n",
       "140   A00711:753:HHCG3DSX7:4:1566:29197:8093   99  chr6  32589832          60   \n",
       "141  A00711:753:HHCG3DSX7:4:1604:27570:18646  163  chr6  32589859          45   \n",
       "142   A00711:753:HHCG3DSX7:4:2118:10872:1188   83  chr6  32589878          60   \n",
       "143  A00711:753:HHCG3DSX7:4:1604:27570:18646   83  chr6  32589997          60   \n",
       "\n",
       "             CIGAR chr_alt   pos_alt insert_size  \\\n",
       "47       25M1I125M       =  32578071         408   \n",
       "49            151M       =  32578028         293   \n",
       "50            151M       =  32577971         204   \n",
       "53   35M2D17M3I96M       =  32578106         226   \n",
       "55       106M3D45M       =  32578034        -226   \n",
       "..             ...     ...       ...         ...   \n",
       "138       80M1I70M       =  32589878         303   \n",
       "140        20S131M       =  32590055         318   \n",
       "141      103M2D48M       =  32589997         288   \n",
       "142       84M2D67M       =  32589728        -303   \n",
       "143       58M1I92M       =  32589859        -288   \n",
       "\n",
       "                                              sequence  \\\n",
       "47   GATAATGGGTTAAAGAAAAATAAAGCCCACTAGATGTAATTTCCTG...   \n",
       "49   ATCTTCACTGAAATACTCATTCCACACATTTTGGTTCTACCTTGTA...   \n",
       "50   GTTCTACCTTGTAATTCCACACCACGAGTCTCATATAAAATGAGAA...   \n",
       "53   TATCAACTTAATATATTTCTATAGGATTTATACTCATAGGGTTTAT...   \n",
       "55   ACAAGTAACATATAACTAAAAATAAGACATGCATAAAAACAAGTTT...   \n",
       "..                                                 ...   \n",
       "138  CTTCAGACACACCATGCTGGAGAAGAGGACAGGACCAGGGGCCAGA...   \n",
       "140  CCACCACTCACATCCCAAACATAGGGAGGAAGTTACTAATTTCTTT...   \n",
       "141  CTCCTGGATTGGGTAATCTCGTGTCGGAGAACCAATCAGCATCTGA...   \n",
       "142  CGTGTCGGAGAACCAATCAGCATCTGAGTTCAACAGCATCATCAGT...   \n",
       "143  AGTACTTGAAAGGTTTGAACCAGTTGCATGTAAAACACTTTAATTG...   \n",
       "\n",
       "                                          base_quality  \n",
       "47   FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFF...  \n",
       "49   FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...  \n",
       "50   FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...  \n",
       "53   FFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFF...  \n",
       "55   FFFFFFFF:FFFFFFFFFFFFFFFFFF,FF,FFFFFFFFFFFF:FF...  \n",
       "..                                                 ...  \n",
       "138  F:FFFFFFFFFFF:,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...  \n",
       "140  FF,FFFF:FFFFFFFF:::FF:FFFFFFFF:FFFFFF:F:FFFFFF...  \n",
       "141  FFFFFFFFFFFFFF::FFFFFFF:FFFFFFFFFFFF,F:FFFFFFF...  \n",
       "142  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...  \n",
       "143  FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...  \n",
       "\n",
       "[82 rows x 11 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = 32570000\n",
    "end = 32590000\n",
    "\n",
    "regstart = hla_gene_information[hla_gene_information['Name'] == f'HLA-{gene}']['Start'].values\n",
    "regend = hla_gene_information[hla_gene_information['Name'] == f'HLA-{gene}']['End'].values\n",
    "strand = hla_gene_information[hla_gene_information['Name'] == f'HLA-{gene}']['Strand'].values\n",
    "\n",
    "reads_apart_max = 1000\n",
    "\n",
    "command = f\"samtools view {bam} chr6:{start}-{end}\"\n",
    "reads = subprocess.run(command, shell = True, capture_output = True, text = True).stdout[:-1].split('\\n')\n",
    "\n",
    "reads = [i.split('\\t') for i in reads if '##' not in i]\n",
    "reads = pd.DataFrame(reads).iloc[:,:11]\n",
    "reads.columns = [\n",
    "    \"ID\", \"flag\", \"chr\", \"pos\", \"map_quality\", \"CIGAR\", \"chr_alt\", \"pos_alt\", \"insert_size\", \"sequence\", \"base_quality\"\n",
    "]\n",
    "\n",
    "reads['pos'] = reads['pos'].astype(int)\n",
    "reads['pos_alt'] = reads['pos_alt'].astype(int)\n",
    "\n",
    "reads = reads[\n",
    "    (reads['chr'] == 'chr6') &\n",
    "    (reads['chr_alt'].isin(['chr6', '='])) &\n",
    "    (reads['pos_alt'] + reads['sequence'].str.len() >= np.ones(len(reads))*(regstart - reads_apart_max)) & \n",
    "    (reads['pos_alt'] <= np.ones(len(reads))*(regend + reads_apart_max))\n",
    "]\n",
    "\n",
    "reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'DRB1'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "\n",
    "ref_samples = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/oneKG.samples', sep = ' ')\n",
    "hlatypes = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/20181129_HLA_types_full_1000_Genomes_Project_panel.txt', sep = '\\t')\n",
    "ref_samples_removed = ref_samples[~ref_samples['SAMPLE'].isin(hlatypes['Sample ID'].tolist())]\n",
    "samples_to_remove = ref_samples_removed['SAMPLE'].tolist()\n",
    "hlatypes = hlatypes[~hlatypes['Sample ID'].isin(samples_to_remove)].sort_values(by = 'Sample ID').reset_index(drop = True)\n",
    "\n",
    "phased_vcf = \"/well/band/users/rbx225/recyclable_files/ref_panels/oneKG/oneKG.chr6.vcf.gz\"\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "\n",
    "# subset_samples = None\n",
    "\n",
    "return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "                            ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "                            hla_gene_information = hla_gene_information,\n",
    "                            hlatypes = hlatypes,\n",
    "                            phased_vcf = phased_vcf, \n",
    "                            reference_allele_ary = reference_allele_ary, \n",
    "                            read_from_QUILT = False, \n",
    "                            subset_vcf_samples = None,\n",
    "                            sample_linker = None)\n",
    "\n",
    "individual = 'NA12878'\n",
    "ix = hlatypes.index[hlatypes['Sample ID'] == 'NA12878'][0]\n",
    "display_indices = np.arange(10)\n",
    "\n",
    "res = visualise_phase(gene, ix, hlatypes, return_dict)\n",
    "compare_phase(display_indices, res)\n",
    "\n",
    "# reference_exclude_samplelist_file = ''\n",
    "# if reference_exclude_samplelist_file != '':\n",
    "#     samples_to_remove = samples_to_remove + lcwgsus.read_tsv_as_lst(reference_exclude_samplelist_file)\n",
    "# if samples_to_remove is not none, save to \"hlauntyped.exclude.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# For GAMCC\n",
    "gene = 'DRB1'\n",
    "ipd_gen_file_dir = '/well/band/users/rbx225/recyclable_files/hla_reference_files/alignments/'\n",
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "samples_to_remove = []\n",
    "\n",
    "gamcc_hla = lcwgsus.read_hla_direct_sequencing(retain = 'fv', unique_two_field = False)\n",
    "gamcc_hla = gamcc_hla[['SampleID', 'Locus', 'Two field1', 'Two field2']].reset_index(drop = True)\n",
    "\n",
    "colnames = ['Sample ID'] + [label for g in HLA_GENES for label in [f'HLA-{g} 1', f'HLA-{g} 2']]\n",
    "hlatypes = pd.DataFrame(columns = colnames)\n",
    "for s in gamcc_hla['SampleID'].unique():\n",
    "    tmp = gamcc_hla[gamcc_hla['SampleID'] == s]\n",
    "    row = [s] + tmp[['Two field1', 'Two field2']].values.ravel().tolist()\n",
    "    hlatypes.loc[len(hlatypes)] = row\n",
    "\n",
    "reference_allele_file = '/well/band/users/rbx225/recyclable_files/hla/b38_reference_alleles.tsv'\n",
    "reference_allele_ary = np.array(lcwgsus.read_tsv_as_lst(reference_allele_file))\n",
    "phased_vcf = \"/well/band/users/rbx225/GAMCC/results/imputation/vcfs/malariaGen_v1_b38/quilt.chr6.vcf.gz\"\n",
    "\n",
    "subset_vcf_samples = lcwgsus.read_tsv_as_lst('/well/band/users/rbx225/GAMCC/data/sample_tsvs/fv_gm_names.tsv')\n",
    "subset_vcf_samples = ','.join(subset_vcf_samples)\n",
    "\n",
    "sample_linker = pd.read_csv('/well/band/users/rbx225/GAMCC/' + SAMPLE_LINKER_FILE)\n",
    "sample_linker = {k:v for k, v in zip(sample_linker['Sample_Name'], sample_linker['Chip_Name'])}\n",
    "\n",
    "return_dict = phase_hla_on_haplotypes(gene = gene, \n",
    "                            ipd_gen_file_dir = ipd_gen_file_dir, \n",
    "                            hla_gene_information = hla_gene_information, \n",
    "                            hlatypes = hlatypes,\n",
    "                            phased_vcf = phased_vcf, \n",
    "                            reference_allele_ary = reference_allele_ary, \n",
    "                            read_from_QUILT = True, \n",
    "                            subset_vcf_samples = subset_vcf_samples,\n",
    "                            sample_linker = sample_linker)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
