{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 100x100 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import io\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import gzip\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import collections\n",
    "sys.path.append('/well/band/users/rbx225/software/lcwgsus/')\n",
    "import lcwgsus\n",
    "from lcwgsus.variables import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Population</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>HLA-A 1</th>\n",
       "      <th>HLA-A 2</th>\n",
       "      <th>HLA-B 1</th>\n",
       "      <th>HLA-B 2</th>\n",
       "      <th>HLA-C 1</th>\n",
       "      <th>HLA-C 2</th>\n",
       "      <th>HLA-DQB1 1</th>\n",
       "      <th>HLA-DQB1 2</th>\n",
       "      <th>HLA-DRB1 1</th>\n",
       "      <th>HLA-DRB1 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>01:01</td>\n",
       "      <td>29:02</td>\n",
       "      <td>08:01</td>\n",
       "      <td>44:03</td>\n",
       "      <td>07:01</td>\n",
       "      <td>16:01</td>\n",
       "      <td>02:01</td>\n",
       "      <td>02:02</td>\n",
       "      <td>03:01</td>\n",
       "      <td>07:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>HG00097</td>\n",
       "      <td>03:01</td>\n",
       "      <td>24:02</td>\n",
       "      <td>07:02</td>\n",
       "      <td>07:02</td>\n",
       "      <td>07:02</td>\n",
       "      <td>07:02</td>\n",
       "      <td>03:01</td>\n",
       "      <td>06:02</td>\n",
       "      <td>13:03</td>\n",
       "      <td>15:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>HG00099</td>\n",
       "      <td>01:01</td>\n",
       "      <td>68:01</td>\n",
       "      <td>08:01</td>\n",
       "      <td>44:02</td>\n",
       "      <td>07:01</td>\n",
       "      <td>07:04</td>\n",
       "      <td>02:01</td>\n",
       "      <td>03:01</td>\n",
       "      <td>03:01</td>\n",
       "      <td>11:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>HG00100</td>\n",
       "      <td>01:01</td>\n",
       "      <td>01:01</td>\n",
       "      <td>08:01</td>\n",
       "      <td>57:01</td>\n",
       "      <td>06:02</td>\n",
       "      <td>07:01</td>\n",
       "      <td>02:01</td>\n",
       "      <td>03:03</td>\n",
       "      <td>03:01</td>\n",
       "      <td>07:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>HG00101</td>\n",
       "      <td>11:01</td>\n",
       "      <td>32:01</td>\n",
       "      <td>27:05</td>\n",
       "      <td>57:01</td>\n",
       "      <td>02:02</td>\n",
       "      <td>06:02</td>\n",
       "      <td>03:02</td>\n",
       "      <td>06:02</td>\n",
       "      <td>04:04</td>\n",
       "      <td>15:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>NA21137</td>\n",
       "      <td>01:01</td>\n",
       "      <td>11:01</td>\n",
       "      <td>57:01</td>\n",
       "      <td>40:06</td>\n",
       "      <td>06:02</td>\n",
       "      <td>15:02</td>\n",
       "      <td>03:03</td>\n",
       "      <td>06:01</td>\n",
       "      <td>09:01</td>\n",
       "      <td>15:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2564</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>NA21141</td>\n",
       "      <td>02:11</td>\n",
       "      <td>24:02</td>\n",
       "      <td>15:05</td>\n",
       "      <td>18:01</td>\n",
       "      <td>03:03</td>\n",
       "      <td>07:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>05:03</td>\n",
       "      <td>07:01</td>\n",
       "      <td>15:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2565</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>NA21142</td>\n",
       "      <td>03:01</td>\n",
       "      <td>24:02</td>\n",
       "      <td>35:01</td>\n",
       "      <td>44:03</td>\n",
       "      <td>04:01</td>\n",
       "      <td>07:06</td>\n",
       "      <td>05:01</td>\n",
       "      <td>06:09</td>\n",
       "      <td>01:01</td>\n",
       "      <td>13:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2566</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>NA21143</td>\n",
       "      <td>11:01</td>\n",
       "      <td>26:01</td>\n",
       "      <td>08:01</td>\n",
       "      <td>57:01</td>\n",
       "      <td>06:02</td>\n",
       "      <td>07:02</td>\n",
       "      <td>02:01</td>\n",
       "      <td>03:03</td>\n",
       "      <td>03:01</td>\n",
       "      <td>07:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2567</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>NA21144</td>\n",
       "      <td>02:01</td>\n",
       "      <td>33:03</td>\n",
       "      <td>57:01</td>\n",
       "      <td>58:01</td>\n",
       "      <td>03:02</td>\n",
       "      <td>06:02</td>\n",
       "      <td>02:01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>03:01</td>\n",
       "      <td>07:01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2568 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Region Population Sample ID HLA-A 1 HLA-A 2 HLA-B 1 HLA-B 2 HLA-C 1  \\\n",
       "0       EUR        GBR   HG00096   01:01   29:02   08:01   44:03   07:01   \n",
       "1       EUR        GBR   HG00097   03:01   24:02   07:02   07:02   07:02   \n",
       "2       EUR        GBR   HG00099   01:01   68:01   08:01   44:02   07:01   \n",
       "3       EUR        GBR   HG00100   01:01   01:01   08:01   57:01   06:02   \n",
       "4       EUR        GBR   HG00101   11:01   32:01   27:05   57:01   02:02   \n",
       "...     ...        ...       ...     ...     ...     ...     ...     ...   \n",
       "2563    SAS        GIH   NA21137   01:01   11:01   57:01   40:06   06:02   \n",
       "2564    SAS        GIH   NA21141   02:11   24:02   15:05   18:01   03:03   \n",
       "2565    SAS        GIH   NA21142   03:01   24:02   35:01   44:03   04:01   \n",
       "2566    SAS        GIH   NA21143   11:01   26:01   08:01   57:01   06:02   \n",
       "2567    SAS        GIH   NA21144   02:01   33:03   57:01   58:01   03:02   \n",
       "\n",
       "     HLA-C 2 HLA-DQB1 1 HLA-DQB1 2 HLA-DRB1 1 HLA-DRB1 2  \n",
       "0      16:01      02:01      02:02      03:01      07:01  \n",
       "1      07:02      03:01      06:02      13:03      15:01  \n",
       "2      07:04      02:01      03:01      03:01      11:01  \n",
       "3      07:01      02:01      03:03      03:01      07:01  \n",
       "4      06:02      03:02      06:02      04:04      15:01  \n",
       "...      ...        ...        ...        ...        ...  \n",
       "2563   15:02      03:03      06:01      09:01      15:01  \n",
       "2564   07:01        NaN      05:03      07:01      15:02  \n",
       "2565   07:06      05:01      06:09      01:01      13:02  \n",
       "2566   07:02      02:01      03:03      03:01      07:01  \n",
       "2567   06:02      02:01        NaN      03:01      07:01  \n",
       "\n",
       "[2568 rows x 13 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hla_gene_information = pd.read_csv('/well/band/users/rbx225/software/QUILT_sus/hla_ancillary_files/hla_gene_information.tsv', sep = ' ')\n",
    "\n",
    "# hlatypes = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/20181129_HLA_types_full_1000_Genomes_Project_panel.txt', sep = '\\t')\n",
    "samples = lcwgsus.read_tsv_as_lst('/well/band/users/rbx225/recyclable_files/ref_panels/oneKG_30x/samples_to_phase.tsv')\n",
    "# hlatypes = hlatypes[hlatypes['Sample ID'].isin(samples)].sort_values(by = 'Sample ID').reset_index(drop = True)\n",
    "# for c in hlatypes.columns[3:]:\n",
    "#     hlatypes[c] = hlatypes[c].astype(str)\n",
    "#     hlatypes[c] = hlatypes[c].str.replace('*', '', regex=False)\n",
    "#     hlatypes[c] = hlatypes[c].apply(lambda x: 'N/A' if '/' in x else x)\n",
    "# hlatypes.to_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/HLA_types_2568_for_phasing.txt', sep = '\\t', header = True, index = False)\n",
    "\n",
    "# The new hlatypes file 1) has all asterisks removed 2) ambiguous types set to N/A 3) only 2568 samples\n",
    "\n",
    "hlatypes = pd.read_csv('/well/band/users/rbx225/GAMCC/results/hla/imputation/ref_panel/auxiliary_files/HLA_types_2568_for_phasing.txt', sep = '\\t')\n",
    "phased_vcf = \"/well/band/users/rbx225/recyclable_files/ref_panels/oneKG_30x/oneKG.chr6.vcf.gz\"\n",
    "\n",
    "start = 25000000\n",
    "end = 34000000\n",
    "\n",
    "command = f\"bcftools view -s {','.join(samples)} -r chr6:\" + str(int(start)) + \"-\" + str(int(end)) + \" \" + phased_vcf\n",
    "vcf = subprocess.run(command, shell = True, capture_output = True, text = True).stdout[:-1].split('\\n')\n",
    "vcf = [i.split('\\t') for i in vcf if '##' not in i]\n",
    "vcf = pd.DataFrame(vcf)\n",
    "vcf.columns = vcf.iloc[0]  # Set the first row as the header\n",
    "vcf = vcf[1:].reset_index(drop = True)\n",
    "vcf['POS'] = vcf['POS'].astype(int)\n",
    "\n",
    "distinct_alleles = {}\n",
    "for g in HLA_GENES:\n",
    "    alleles = np.append(hlatypes[f'HLA-{g} 1'].astype(str).to_numpy(), hlatypes[f'HLA-{g} 2'].astype(str).to_numpy())\n",
    "    alleles = np.unique(alleles)\n",
    "    alleles = alleles[alleles != 'nan']\n",
    "    distinct_alleles[g] = alleles\n",
    "\n",
    "test = vcf.copy()\n",
    "\n",
    "for g in HLA_GENES:\n",
    "    r = hla_gene_information[hla_gene_information['Name'] == f'HLA-{g}']\n",
    "    start = r['Start'].values[0]\n",
    "    all_snp_pos = vcf['POS'].tolist()\n",
    "    position = start\n",
    "    n_alleles = 0\n",
    "\n",
    "    while n_alleles < len(distinct_alleles[g]):\n",
    "        if position in all_snp_pos:\n",
    "            pass\n",
    "        else:\n",
    "            twofield = distinct_alleles[g][n_alleles]\n",
    "            common_cols = ['chr6', position, f'HLA_{g}*{twofield}', 'T', 'A', '.', 'PASS', '.', 'GT']\n",
    "            allele1 = (hlatypes[f'HLA-{g} 1'] == twofield).astype(int).values\n",
    "            allele2 = (hlatypes[f'HLA-{g} 2'] == twofield).astype(int).values\n",
    "\n",
    "            gts = [f'{a1}/{a2}' for a1, a2 in zip(allele1, allele2)]\n",
    "            vcf.loc[len(vcf)] = common_cols + gts\n",
    "            n_alleles += 1\n",
    "        position += 1\n",
    "\n",
    "vcf = vcf.sort_values(by = 'POS', ascending = True).reset_index(drop = True)\n",
    "\n",
    "# metadata = lcwgsus.read_metadata(phased_vcf, filetype = 'gzip', comment = '#', new_cols = None)\n",
    "# metadata[-1] = '\\t'.join(vcf.columns)\n",
    "# lcwgsus.save_vcf(vcf,\n",
    "#              metadata,\n",
    "#              rezip = True,\n",
    "#              prefix='chr',\n",
    "#              outdir='/well/band/users/rbx225/GAMCC/results/phasing/HLA_1KG_BEAGLE/',\n",
    "#              save_name='unphased.1KG.chr6.vcf.gz'\n",
    "#              )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "beagle_phasing = \"/well/band/users/rbx225/GAMCC/results/phasing/HLA_1KG_BEAGLE/phased.1KG.chr6.vcf.gz\"\n",
    "sus_phasing = \"/well/band/users/rbx225/GAMCC/results/hla_ref_panel/oneKG_mGenv1/oneKG_mGenv1_HLA_calls.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
